test/integration/etcd/restart_server.go

```diff
@@ -0,0 +1,496 @@
+	if err != nil {
+		t.Fatalf("create temp file failed: %v", err)
+	}
+	if err = os.WriteFile(saSigningKeyFile.Name(), []byte(ecdsaPrivateKey), 0666); err != nil {
+		t.Fatalf("write file %s failed: %v", saSigningKeyFile.Name(), err)
+	}
+
+	kubeAPIServerOptions := options.NewServerRunOptions()
+	kubeAPIServerOptions.SecureServing.BindPort = 0
+	kubeAPIServerOptions.ServiceAccountSigningKeyFile = saSigningKeyFile.Name()
+	kubeAPIServerOptions.Etcd.StorageConfig.Transport.ServerList = []string{framework.GetEtcdURL()}
+	kubeAPIServerOptions.ServiceClusterIPRanges = defaultServiceClusterIPRange.String()
+	kubeAPIServerOptions.Authentication.ServiceAccounts.Issuers = []string{"https://foo.bar.example.com"}
+	kubeAPIServerOptions.Authentication.ServiceAccounts.KeyFiles = []string{saSigningKeyFile.Name()}
+	kubeAPIServerOptions.APIEnablement.RuntimeConfig["api/all"] = "true"
+	// we make requests to all resources, don't log warnings about deprecated ones
+	restclient.SetDefaultWarningHandler(restclient.NoWarnings{})
+	caData, caFilename, caCert, caKey := writeCACertFiles(t, certDir)
+	clientCertData, clientKeyData := writeClientCerts(t, caCert, caKey, certDir, 10_000*time.Hour)
+	servingCertFilename, servingKeyFilename := writeServingCerts(t, caCert, caKey, certDir, 10_000*time.Hour)
+	// do not need to set CertDirectory since we are providing our own certs
+	kubeAPIServerOptions.Authentication.ClientCert.ClientCA = caFilename
+	kubeAPIServerOptions.SecureServing.ServerCert.CertKey.KeyFile = servingKeyFilename
+	kubeAPIServerOptions.SecureServing.ServerCert.CertKey.CertFile = servingCertFilename
+
+	rawClient, kvClient, err := integration.GetEtcdClients(kubeAPIServerOptions.Etcd.StorageConfig.Transport)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() { rawClient.Close() })
+	// make sure we start with a clean slate
+	if _, err := kvClient.Delete(context.Background(), "/registry/", clientv3.WithPrefix()); err != nil {
+		t.Fatal(err)
+	}
+
+	stopCh := make(chan struct{})
+	errCh := make(chan error, 100)
+
+	cleanup, err := start(t, stopCh, errCh, kubeAPIServerOptions, nil, customFlags, saSigningKeyFile, certDir, rawClient)
+	if err != nil {
+		if rawClient != nil {
+			rawClient.Close()
+		}
+		cleanup()
+		return nil, err
+	}
+
+	// need to do this AFTER start() to get the port
+	address := fmt.Sprintf("127.0.0.1:%d", kubeAPIServerOptions.SecureServing.BindPort)
+
+	kubeClientConfig := &restclient.Config{
+		Host: address,
+		TLSClientConfig: restclient.TLSClientConfig{
+			ServerName: "kubernetes.default.svc",
+			CAData:     caData,
+			CertData:   clientCertData,
+			KeyData:    clientKeyData,
+		},
+		QPS:   99999,
+		Burst: 9999,
+	}
+
+	kubeClient := clientset.NewForConfigOrDie(kubeClientConfig)
+	klog.Infof("kubeClientConfig.Host: %v", kubeClientConfig.Host)
+
+	err = healthCheck(t, kubeClient, errCh)
+	if err != nil {
+		if rawClient != nil {
+			rawClient.Close()
+		}
+		cleanup()
```

Comment:
Would this run in case of a panic? Wondering if we need to put this in a `defer` so they always run?

Reply:
to clarify, do you mean something like this?
```go
if err != nil {
  defer cleanup()
  if rawClient != nil {
      rawClient.Close()
  }
  return nil, err
}
```

Reply:
Added https://github.com/kubernetes/kubernetes/pull/119050/commits/4b636efc73a3b8529f009e05d5bd1487b885aea4#diff-78f66df2cd0735f235d73b6d4dbf70b4ab68da8298049f3ac1c7d0b6f0f92000R414-R423